### PR TITLE
Clear the mempool by using a sync method

### PR DESCRIPTION
**Pull request by upbqdn**
Addresses https://github.com/ZcashFoundation/zebra/pull/2773#discussion_r711861630

I changed https://github.com/ZcashFoundation/zebra/blob/44ac06775bf6bb77dea06d53a6fc5e04a79baf31/zebra-state/src/service/chain_tip.rs#L313 to a private method so that `get_tip_change()` doesn't miss any tip changes.

### CODE REVIEW

File: zebra-state/src/service/chain_tip.rs
Comment (teor2345) [inline zebra-state/src/service/chain_tip.rs]: ### Tricky tokio edge case

We'll need to disable tokio's cooperative scheduling, otherwise we will get spurious `None`s from `now_or_never`.

(The warning about starvation doesn't apply, since we're using `now_or_never`, which always returns immediately.)

https://docs.rs/tokio/1.11.0/tokio/task/index.html#cooperative-scheduling

I'll open a PR to update the [async RFC](https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0011-async-rust-in-zebra.md). This seems like an edge case we'll need to remember.

```suggestion
        match tokio::task::unconstrained(self.tip_change()).now_or_never().transpose() {
```

File: zebra-state/src/service/chain_tip.rs
Comment (teor2345) [inline zebra-state/src/service/chain_tip.rs]: ### Tricky shutdown edge case

If `tip_block_change` returns an error, that means the `watch::Sender` has been dropped, so we won't get any more state changes:
https://github.com/ZcashFoundation/zebra/blob/957e12e4ca28438ca2c64ca4a1870210f2b3e9b7/zebra-state/src/service/chain_tip.rs#L377-L380

I think we'll want to return that error here, so we can return an error from `Mempool::poll_ready`.
(This error should only happen when Zebra is shutting down.)

### Naming suggestion

Rust APIs typically don't use `get`: 
https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

There might be a better name here, but I can't think of it right now. (Names are hard!)

### Documentation suggestion

It's good to avoid duplicating the detailed documentation on `wait_for_tip_change`.
But let's point to it, so people know where to look for details.

```suggestion
    ///
    /// See [`wait_for_tip_change`] for details.
    pub fn last_tip_change(&mut self) -> Result<Option<TipAction>, watch::error::RecvError> {
```

File: zebra-state/src/service/chain_tip.rs
Comment (teor2345) [inline zebra-state/src/service/chain_tip.rs]: ### Other PRs

I'll need this method in #2729, because block gossip is an async task that waits for state changes.

But we should definitely rename it to avoid confusion.

```suggestion
    pub async fn wait_for_tip_change(&mut self) -> Result<TipAction, watch::error::RecvError> {
```

**Review by teor2345 — COMMENTED** [review]
This looks good, there are just a few tricky edge cases we need to handle.

File: zebra-state/src/service/chain_tip.rs
Comment (upbqdn) [reply zebra-state/src/service/chain_tip.rs]: Thanks for the warning about duplicating the docs. I was thinking it might not have been right what I did. Also, thanks for the note about `get`. Names consume a substantial amount of time in my workflow. What I sometimes do is that I name a particular thing `foo` and then rename it when the code is finished. This "lazy-evaluation" approach sometimes works.

Comment (teor2345) [reply zebra-state/src/service/chain_tip.rs]: It's ok to take shortcuts to get something working, then ask the reviewer for help improving things.
(Some shortcuts are also totally ok to commit into the `main` branch.)

Comment (teor2345) [reply zebra-state/src/service/chain_tip.rs]: Also I think what you've done with the documentation is fine here.

File: zebra-state/src/service/chain_tip.rs
Comment (teor2345) [reply zebra-state/src/service/chain_tip.rs]: I just found out that `tokio::task::unconstrained` is new in tokio 1.12, but we're only using tokio 0.3.7.

Here's the documentation for 0.3.7 - there's no `unconstrained`, but there is preemption:
https://docs.rs/tokio/0.3.7/tokio/task/index.html#what-are-tasks
https://tokio.rs/blog/2020-04-preemption

So I think it might be better to call `ChainTipChange::action` directly from `last_tip_change`.

You can get the last block using [watch::Receiver::borrow](https://docs.rs/tokio/0.3.7/tokio/sync/watch/struct.Receiver.html#method.borrow).

Here are some other things to be aware of during the refactor.
(They're not documented in the code - sorry about that!)
* the `assert!` only applies to `wait_for_tip_change` - if the last hash is the same, that's when `last_tip_change` will want to return `None`
* the `last_change_hash` update needs to be done for both `wait_for_tip_change` and `last_tip_change`

Also, the code will be simpler and faster, which is nice.

File: zebra-state/src/service/chain_tip.rs
Comment (teor2345) [reply zebra-state/src/service/chain_tip.rs]: Note: If we call `ChainTipChange::action` directly, the error here can't happen any more.

**Review by teor2345 — COMMENTED** [review]
Seems like I gave some bad advice here - we can't use `unconstrained` until we upgrade tokio.

But the alternative refactor gives us simpler and faster code, so let's just do that.

**Comment by upbqdn** [comment]
All suggestions should be implemented now.

File: zebra-state/src/service/chain_tip.rs
Comment (upbqdn) [reply zebra-state/src/service/chain_tip.rs]: > Note: If we call ChainTipChange::action directly, the error here can't happen any more.

I should have read the note more properly. I was still trying to return `Result` from `last_tip_change`, and I couldn't figure out how to return a `RecvError`. After a while, I realized I can just return `Option<TipAction>`.

File: zebra-state/src/service/chain_tip.rs
Comment (jvff) [inline zebra-state/src/service/chain_tip.rs]: Optional suggestion: You can remove two levels of indentation by using the try operator (`?`), but that would require adding an extra binding:

```suggestion
        let block_guard = self.receiver.borrow();
        let block = block_guard.as_ref()?;
        
        // Ignore an unchanged tip.
        if Some(block.hash) == self.last_change_hash {
            return None;
        }

        let action = self.action(block.clone());

        self.last_change_hash = Some(block.hash);

        Some(action)
```

File: zebrad/src/components/mempool.rs
Comment (jvff) [inline zebrad/src/components/mempool.rs]: Optional suggestion: you could simplify this by ignoring all fields like so:

```suggestion
        if let Some(TipAction::Reset { .. }) = self.chain_tip_change.last_tip_change() {
```

**Review by jvff — APPROVED** [review]
LGTM. I added a few optional suggestions, but they're minor.

File: zebra-state/src/service/chain_tip.rs
Comment (upbqdn) [reply zebra-state/src/service/chain_tip.rs]: I was trying the `?` operator, but I couldn't get it to work. Thanks.

File: zebrad/src/components/mempool.rs
Comment (upbqdn) [reply zebrad/src/components/mempool.rs]: I was just about to push this. I saw it in Teor's PR today.

File: zebra-state/src/service/chain_tip.rs
Comment (jvff) [reply zebra-state/src/service/chain_tip.rs]: No problem :D Yeah, the tricky part is satisfying the borrow checker. `self.receiver.borrow()` returns a wrapper type that `Deref`s into its inner type (i.e., `Option<_>`), so you can call `.as_ref()` directly. The problem is that chaining calls makes Rust drop intermediate types as soon as possible, so the wrapper is dropped because it's not stored anywhere. The chain of events is (at least AFAICT):

1. The wrapper type is obtained and stored in a temporary;
2. That temporary is then borrowed and dereferenced to get a borrow to the inner type, i.e. `&Option<T>`;
3. That borrow is used to call `fn as_ref(&self) -> Option<&T>`, so now we have a result that depends on the borrow;
4. The wrapper type is dropped;
5. The result is stored.

The problem is in steps 4 and 5, because we still have inside the result a borrow to the temporary, even though it was dropped. So the Rust compiler fails to compile the code. By creating the `blocks_guard` binding we make sure that it is only dropped at the end of the scope.

**Review by jvff — APPROVED** [review]

**Comment by upbqdn** [comment]
Now I'm not sure whether to merge this into `main` or the parent branch.

File: zebra-state/src/service/tests.rs
Comment (teor2345) [inline zebra-state/src/service/tests.rs:303]: Nice catch!

Thanks for refactoring these tests.

File: zebra-state/src/service/chain_tip.rs
Comment (teor2345) [inline zebra-state/src/service/chain_tip.rs]: Performance Nit: The borrow guard blocks the `watch::Sender` until the end of this method. That might slow down block commits in the state.

`last_tip_change` and `action` hold the lock longer than anywhere else. They doesn't do anything expensive yet, but that could change. 

So we're probably best to just clone immediately, and drop the lock. (`best_tip_block` does a clone, which is cheap. `ChainTipBlock::clone` only duplicates ~90 bytes plus an `Arc` atomic increment.)

> Outstanding borrows hold a read lock. This means that long lived borrows could cause the send half to block. It is recommended to keep the borrow as short lived as possible.

https://docs.rs/tokio/0.3.7/tokio/sync/watch/struct.Receiver.html#method.borrow

```suggestion
        let block = self.best_tip_block()?;
```

**Review by teor2345 — COMMENTED** [review]
We could merge this as-is, but I'd like to fix a potential future performance issue.
(I expect we'll be using `last_tip_change` a lot, at the same time as the state is committing thousands of blocks.)

**Comment by teor2345** [comment]
> Now I'm not sure whether to merge this into `main` or the parent branch.

We could:
* merge #2777 into #2773
* merge #2773 into `main`
* rebase #2764 on the new `main`

But I'm happy to let @conradoplg make the final call here, because #2764 is his PR.

File: zebra-state/src/service/chain_tip.rs
Comment (teor2345) [reply zebra-state/src/service/chain_tip.rs]: Please feel free to merge as soon as this is fixed - I don't need to review again.

**Comment by conradoplg** [comment]
> But I'm happy to let @conradoplg make the final call here, because #2764 is his PR.

I agree with the plan :+1: I can merge this after that last required change

**Comment by upbqdn** [comment]
@conradoplg both PRs should be ready.

File: zebra-state/src/service/chain_tip.rs
Comment (upbqdn) [reply zebra-state/src/service/chain_tip.rs]: Thanks for the explanation!

**Pull request merged**
